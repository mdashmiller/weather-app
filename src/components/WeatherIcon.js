import React from 'react'
import Icon from './Icon'

const WeatherIcon = ({ id, daytime }) => {
	// determines whether to display a
	// cloud or sun icon dependent upon
	// weather conditions

	// weather-dependent icon details
	let celestialBody
	if (daytime) {
		celestialBody = "fas fa-sun fa-3x"
	} else {
		celestialBody = "fas fa-moon fa-3x"
	}
	const cloud = "fas fa-cloud fa-3x"	

	if (
			(id > 959 && id < 963) ||
			(id > 899 && id < 903) ||
			id === 804 ||
			id === 803 ||
			(id > 700 && id < 782) ||
			(id > 599 && id < 623) ||
			(id > 499 && id < 532) ||
			(id > 299 && id < 322) ||
			(id > 199 && id < 233)
	) {
		return <Icon dark className={cloud}></Icon>
	}
	return <Icon className={celestialBody}></Icon>
}
	
export default WeatherIcon
