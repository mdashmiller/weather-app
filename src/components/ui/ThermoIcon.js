import React from 'react'
import Icon from './Icon'

// weather-dependent icon details
const thermoHot = "fas fa-thermometer-full fa-3x"
const thermoWarm = "fas fa-thermometer-three-quarters fa-3x"
const thermoCool = "fas fa-thermometer-half fa-3x"
const thermoCold = "fas fa-thermometer-empty fa-3x"

const ThermoIcon = ({ temp }) => {
	// determines which thermometer icon to
	// display dependent upon temperature
	if (temp > 32 && temp < 69) {
		return <Icon className={thermoCool}></Icon>
	} else if (temp >= 69 && temp < 86) {
		return <Icon className={thermoWarm}></Icon>
	} else if (temp >= 86) {
		return <Icon className={thermoHot}></Icon>
	} else {
		return <Icon className={thermoCold}></Icon>
	}
}

export default ThermoIcon
