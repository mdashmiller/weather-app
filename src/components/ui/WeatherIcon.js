import React from 'react'
import Icon from './Icon'
import PropTypes from 'prop-types'

const WeatherIcon = ({ code, day }) => {
	// determines whether to display a
	// cloud or sun icon dependent upon
	// weather conditions

	// weather-dependent icon details
	const celestialBody = day
		? "fas fa-sun fa-3x" : "fas fa-moon fa-3x"
	const cloud = "fas fa-cloud fa-3x"	

	if (
			(code > 959 && code < 963) ||
			(code > 899 && code < 903) ||
			code === 804 ||
			code === 803 ||
			(code > 700 && code < 782) ||
			(code > 599 && code < 623) ||
			(code > 499 && code < 532) ||
			(code > 299 && code < 322) ||
			(code > 199 && code < 233)
	) {
		return <Icon dark className={cloud}></Icon>
	}
	return <Icon className={celestialBody}></Icon>
}

WeatherIcon.propTypes = {
	code: PropTypes.number,
	day: PropTypes.bool
}

export default WeatherIcon
