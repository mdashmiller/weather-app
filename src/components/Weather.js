import React, { Component } from 'react'
import WeatherIcon from './WeatherIcon'
import ThermoIcon from './ThermoIcon'

class Weather extends Component {

	state = {
		day: true,
		place: '',
		temp: '',
		description: '', 
		codition: ''
	}

	// component methods

	setWeatherInfo = result => {
		// takes results from API call and
		// routes data to the proper channels
		this.setTemp(result.main.temp)
		this.setDescription(result.weather[0].description)
		this.setPlace(result.name)
		this.setCondition(result.weather[0].id)
		this.dayOrNight(result.dt, result.sys.sunrise, result.sys.sunset)
	}

	setTemp = tempK => {
		// sets this.state.temp to the result
		// obtained from Open Weather Map
		const tempF =  this.convertToFahrenheit(tempK)
		this.setState( {temp: tempF} )
	}

	convertToFahrenheit = k => {
		// takes a temperature in Kelvin
		// and returns it in Fahrenheit
		const floatK = parseFloat(k)
		return (
			(1.8 * (floatK - 273)) + 32
		).toFixed(1)
	}

	setDescription = weather =>
		// sets this.state.description to the result
		// obtained from Open Weather Map
		this.setState( {description: weather })

	setPlace = place =>
		// sets this.state.name to the result
		// obtained from Open Weather Map
		this.setState( {place} )

	setCondition = id =>
		// sets this.state.condition to the result
		// obtained from Open Weather Map
		this.setState( {condition: id} )

	dayOrNight = (time, sunrise, sunset) => {
		// determines if it is day or night
		// and sets state accordingly
		if (time >= sunrise && time < sunset) {
			this.setState({ day: true })
		} else {
			this.setState({ day: false })
		}
	}

	// lifecycle hooks

	componentDidMount () {
		this.setWeatherInfo(this.props.result)
	}

	render () {
		const {
			day,
			place,
			temp,
			description,
			condition
		} = this.state
		return (
			<div>
				<h1
					className={`${day ? 'day-title' : 'night-title'}`}
				>
					{place}
				</h1>
				<h2>{temp} F</h2>
				<h3>{description}</h3>
				<WeatherIcon 
					code={condition} 
					day={day}
				/>
				<ThermoIcon temp={temp} />
			</div>
		)
	}
}

export default Weather
