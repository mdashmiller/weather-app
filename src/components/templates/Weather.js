import React, { Component } from 'react'
import WeatherIcon from '../ui/WeatherIcon'
import ThermoIcon from '../ui/ThermoIcon'
import PropTypes from 'prop-types'

class Weather extends Component {

	state = {
		day: null,
		place: '',
		temp: '',
		description: '', 
		codition: null
	}

	// component methods

	setWeatherInfo = result => {
		// takes results from API call and
		// routes data to the proper channels
		this.setTemp(result.main.temp)
		this.setDescription(result.weather[0].description)
		this.setPlace(result.name)
		this.setCondition(result.weather[0].id)
		this.dayOrNight(result.dt, result.sys.sunrise, result.sys.sunset)
	}

	setTemp = tempK => {
		// sets this.state.temp to the result
		// obtained from Open Weather Map
		const tempF =  this.convertToFahrenheit(tempK)
		this.setState( {temp: tempF} )
	}

	convertToFahrenheit = k => {
		// takes a temperature in Kelvin
		// and returns it in Fahrenheit
		return (
			(1.8 * (k - 273)) + 32
		).toFixed(1)
	}

	setDescription = weather =>
		// sets this.state.description to the result
		// obtained from Open Weather Map
		this.setState( {description: weather })

	setPlace = place =>
		// sets this.state.name to the result
		// obtained from Open Weather Map
		this.setState( {place} )

	setCondition = id =>
		// sets this.state.condition to the result
		// obtained from Open Weather Map
		this.setState( {condition: id} )

	dayOrNight = (time, sunrise, sunset) => {
		// determines if it is day or night
		// and sets state accordingly
		if (time >= sunrise && time < sunset) {
			this.setState({ day: true })
		} else {
			this.setState({ day: false })
		}
	}

	// lifecycle hooks

	componentDidMount () {
		this.setWeatherInfo(this.props.result)
	}

	render () {
		const {
			day,
			place,
			temp,
			description,
			condition
		} = this.state
		return (
			<div>
				<h1 className="title-gold">
					{place}
				</h1>
				<h2>{temp} F</h2>
				<h3>{description}</h3>
				<WeatherIcon 
					code={condition} 
					day={day}
				/>
				<ThermoIcon temp={temp} />
			</div>
		)
	}
}

Weather.propTypes = {
	result: PropTypes.shape({
		name: PropTypes.string,
		main: PropTypes.shape({
			temp: PropTypes.number
		}),
		weather: PropTypes.arrayOf(
			PropTypes.shape({
				description: PropTypes.string,
				id: PropTypes.number
			})
		),
		dt: PropTypes.number,
		sys: PropTypes.shape({
			sunrise: PropTypes.number,
			sunset: PropTypes.number
		})
	})
}

export default Weather
