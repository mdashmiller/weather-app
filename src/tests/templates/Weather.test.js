import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import Weather from '../../components/templates/Weather'
import WeatherIcon from '../../components/ui/WeatherIcon'
import ThermoIcon from '../../components/ui/ThermoIcon'

Enzyme.configure({ adapter: new Adapter() })

// fake weather API data
const result = {
	name: 'Testville',
	main: {
		temp: 297
	},
	weather: [
		{
			description: 'all good!',
			id: 100
		}
	],
	dt: 2,
	sys: {
		sunrise: 1,
		sunset: 3
	}
}

describe('<Weather /> rendering', () => {

	it('renders without crashing', () => {
	  	const div = document.createElement('div')
	  	ReactDOM.render(<Weather result={result} />, div)
	  	ReactDOM.unmountComponentAtNode(div)
	})

	test('has a valid snapshot', () => {
		const component = renderer.create(<Weather result={result} />)
		let tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('renders 1 <div> element', () => {
		const wrapper = shallow(<Weather result={result} />)
		expect(wrapper.find('div').length).toBe(1)
	})

	it('renders 1 <h1> element', () => {
		const wrapper = shallow(<Weather result={result} />)
		expect(wrapper.find('h1').length).toBe(1)
	})

	it('renders 1 <h2> element', () => {
		const wrapper = shallow(<Weather result={result} />)
		expect(wrapper.find('h2').length).toBe(1)
	})

	it('renders 1 <h3> element', () => {
		const wrapper = shallow(<Weather result={result} />)
		expect(wrapper.find('h3').length).toBe(1)
	})

	it('renders 1 <WeatherIcon /> component', () => {
		const wrapper = shallow(<Weather result={result} />)
		expect(wrapper.find('WeatherIcon').length).toBe(1)
	})

	it('renders 1 <ThermoIcon /> component', () => {
		const wrapper = shallow(<Weather result={result} />)
		expect(wrapper.find('ThermoIcon').length).toBe(1)
	})

})

describe('directly invoking setWeatherInfo()', () => {

	it('calls the proper functions', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()

		jest.spyOn(instance, 'setTemp')
		jest.spyOn(instance, 'setDescription')
		jest.spyOn(instance, 'setPlace')
		jest.spyOn(instance, 'setCondition')
		jest.spyOn(instance, 'dayOrNight')

		instance.setWeatherInfo(result)
		expect(instance.setTemp).toHaveBeenCalledWith(297)
		expect(instance.setDescription).toHaveBeenCalledWith('all good!')
		expect(instance.setPlace).toHaveBeenCalledWith('Testville')
		expect(instance.setCondition).toHaveBeenCalledWith(100)
		expect(instance.dayOrNight).toHaveBeenCalledWith(2, 1, 3)
	})

})

describe('directly invoking setTemp()', () => {

	it('calls convertToFahrenheit()', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()
		jest.spyOn(instance, 'convertToFahrenheit')
		instance.setTemp(result.main.temp)
		expect(instance.convertToFahrenheit).toHaveBeenCalledWith(297)
	})	

	it('sets state with the proper value for temp', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()
		instance.setTemp(result.main.temp)
		expect(wrapper.state('temp')).toBe('75.2')
	})

})

describe('directly invoking convertToFahrenheit()', () => {

	it('properly converts Kelvin to Fahrenheit', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()
		jest.spyOn(instance, 'convertToFahrenheit')
		instance.convertToFahrenheit(0)
		expect(instance.convertToFahrenheit).toHaveReturnedWith('-459.4')
	})

})

describe('directly invoking setDescription()', () => {

	it('sets description property in state to result.weather[0].description', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()
		instance.setDescription(result.weather[0].description)
		expect(wrapper.state('description')).toBe('all good!')
	})

})

describe('directly invoking setPlace()', () => {

	it('sets place property in state to result.name', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()
		instance.setPlace(result.name)
		expect(wrapper.state('place')).toBe('Testville')
	})

})

describe('directly invoking setCondition()', () => {

	it('sets condition property in state to result.weather[0].id', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()
		instance.setCondition(result.weather[0].id)
		expect(wrapper.state('condition')).toBe(100)
	})

})

describe('directly invoking dayOrNight()', () => {

	it('sets day property in state in relation to sunrise and sunset', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()
		// time between sunrise and sunset
		instance.dayOrNight(2, 1, 3)
		expect(wrapper.state('day')).toBe(true)
		// time is after sunset
		instance.dayOrNight(3, 1, 2)
		expect(wrapper.state('day')).toBe(false)
		// time is equal to sunrise
		instance.dayOrNight(1, 1, 2)
		expect(wrapper.state('day')).toBe(true)
		// time is equal to sunset
		instance.dayOrNight(2, 1, 2)
		expect(wrapper.state('day')).toBe(false)
	})

})

describe('directly invoking componentDidMount()', () => {

	it('calls setWeatherInfo()', () => {
		const wrapper = shallow(<Weather result={result} />)
		const instance = wrapper.instance()
		jest.spyOn(instance, 'setWeatherInfo')
		instance.componentDidMount()
		expect(instance.setWeatherInfo).toHaveBeenCalledWith(result)
	})

})
