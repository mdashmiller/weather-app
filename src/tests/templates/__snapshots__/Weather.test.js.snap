// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather /> rendering has a valid snapshot 1`] = `
<div>
  <h1
    className="title-gold"
  >
    Testville
  </h1>
  <h2>
    75.2
     F
  </h2>
  <h3>
    all good!
  </h3>
  <i
    className="fas fa-sun fa-3x sc-bdVaJa bSIgLK"
  />
  <i
    className="fas fa-thermometer-three-quarters fa-3x sc-bdVaJa bSIgLK"
  />
</div>
`;
